stablehlo/reference/Tensor.cpp

```diff
@@ -52,65 +52,57 @@ int64_t getSizeInBytes(Type type) {
 
 }  // namespace
 
-namespace detail {
-
-Buffer::Buffer(ShapedType type) : type_(type), data_(getSizeInBytes(type), 0) {}
-
-Buffer::Buffer(ShapedType type, void *data)
-    : Buffer(type, static_cast<const void *>(data)) {}
+Tensor::Tensor() {}
 
-Buffer::Buffer(ShapedType type, const void *data)
-    : type_(type),
-      data_(static_cast<const char *>(data),
-            static_cast<const char *>(data) + getSizeInBytes(type)) {}
+Tensor::Tensor(ShapedType type) : type_(type) {
+  blob_ = HeapAsmResourceBlob::allocateAndCopy(
+      ArrayRef<char>(std::vector<char>(getSizeInBytes(type), 0)),
+      alignof(char));
+}
 
-}  // namespace detail
+Tensor::Tensor(ShapedType type, AsmResourceBlob blob)
+    : type_(type), blob_(std::move(blob)) {}
 
-Tensor::Tensor() {}
-
-Tensor::Tensor(ShapedType type)
-    : impl_(llvm::makeIntrusiveRefCnt<detail::Buffer>(type)) {}
-
-Tensor::Tensor(DenseElementsAttr attr) {
-  // TODO(sdasgup3): We're using DenseElementsAttr::getRawData() here for
-  // simplicity, because it provides a contiguous representation of underlying
-  // data in most cases. However, this doesn't always work (e.g. for splat or
-  // for i1), so we'll be migrating to something more reliable in the near
-  // future.
-  impl_ = llvm::makeIntrusiveRefCnt<detail::Buffer>(attr.getType(),
-                                                    attr.getRawData().data());
+Tensor::Tensor(const Tensor &other) {
+  type_ = other.getType();
+  blob_ = HeapAsmResourceBlob::allocateAndCopy(other.getData(), alignof(char));
```

Comment:
In the current design, `Tensor::Tensor(const Tensor &other)` and `Tensor &Tensor::operator=(const Tensor &other)` don't involve copying memory ([see details](https://github.com/llvm/llvm-project/blob/dfbaf90043841d18f5871e12b10c674b9b3abf99/llvm/include/llvm/ADT/IntrusiveRefCntPtr.h#L174-L196)), but it looks like in the proposed design, they both involve copying. Do we consider that to be a downside? If yes, is there something that can be done about it?

Reply:
Thanks for catching this. This allows lots of unnecessary copies. Fixed it by reference counting the internal storage.